src/core/core/logging.cpp

```diff
@@ -148,10 +134,26 @@ DSN_API void dsn_log(const char *file,
                      dsn_log_level_t log_level,
                      const char *str)
 {
-    ::dsn::logging_provider *logger = ::dsn::service_engine::instance().logging();
-    if (logger != nullptr) {
-        logger->dsn_log(file, function, line, log_level, str);
-    } else {
-        printf("%s:%d:%s():%s\n", file, line, function, str);
-    }
+    dsn::logging_provider *logger = dsn::logging_provider::instance();
+    logger->dsn_log(file, function, line, log_level, str);
 }
+
+namespace dsn {
+
+std::unique_ptr<logging_provider> logging_provider::_logger =
+    std::unique_ptr<logging_provider>(nullptr);
+
+logging_provider *logging_provider::instance()
+{
+    static std::unique_ptr<logging_provider> default_logger =
```

Comment:
Why `default_logger` defined as static? its lifetime will be kept until process exit.

Reply:
because if not use static, a new logger will be created by each call to logging_provider::instance().
Yet its lifetime will be kept untiil process exit, but there is extra one static logger only, so I think it`s ok.

Reply:
It's because that `logging_provider::instance()` may be called befaore `_logger ` has been created, we have to use `default_logger` instead that time?

Reply:
Yes. `_logger` is the user-customized logger. For now, dsn_runtime is the "user", which reads the configuration and `set_logger` to replace default_logger (`dsn_log_init`). But if an application (like unit-test) without binding the runtime wants to print log, it doesn't have to set a logger of its own.